docs/contributing.rst

```diff
@@ -0,0 +1,261 @@
+.. _contributing:
+
+Contributing
+------------
+
+Thank you for your interest in contributing! We welcome all contributions
+no matter their size. Please read along to learn how to get started. If you
+get stuck, feel free to reach for help in our
+`Gitter channel <https://gitter.im/ethereum/web3.py>`_.
+
+Setting the stage
+~~~~~~~~~~~~~~~~~
+
+First, you need to clone the repository. Web3.py depends on submodules, so you
+need to clone the repo with the ``--recursive`` flag. Example:
+
+.. code:: sh
+
+    $ git clone --recursive https://github.com/ethereum/web3.py.git
+    $ cd web3.py
+
+
+Install all development dependencies:
+
+.. code:: sh
+
+    $ pip install -e .[dev]
+
+
+Using Docker
+~~~~~~~~~~~~
+
+Developing within Docker is not required, but if you prefer that workflow, use
+the *sandbox* container provided in the **docker-compose.yml** file.
+
+To start up the test environment, run:
+
+.. code:: sh
+
+    $ docker-compose up -d
+
+
+This will build a Docker container set up with an environment to run the
+Python test code.
+
+.. note::
+
+    This container does not have `go-ethereum` installed, so you cannot run
+    the go-ethereum test suite.
+
+To run the Python tests from your local machine:
+
+.. code:: sh
+
+    $ docker-compose exec sandbox bash -c 'pytest -n 4 -f -k "not goethereum"'
+
+
+You can run arbitrary commands inside the Docker container by using the
+`bash -c` prefix.
+
+.. code:: sh
+
+    $ docker-compose exec sandbox bash -c ''
+
+
+Or, if you would like to open a session to the container, run:
+
+.. code:: sh
+
+    $ docker-compose exec sandbox bash
+
+
+Running the tests
+~~~~~~~~~~~~~~~~~
+
+A great way to explore the code base is to run the tests.
+
+You can run all tests with:
+
+.. code:: sh
+
+    $ pytest
+
+
+However, running the entire test suite takes a very long time and is generally impractical.
+Typically, you'll just want to run a subset instead, like:
+
+.. code:: sh
+
+    $ pytest tests/core/eth-module/test_accounts.py
+
+
+You can use ``tox`` to run all the tests for a given version of Python:
+
+.. code:: sh
+
+   $ tox -e py37-core
+
+
+Linting is also performed by the CI. You can save yourself some time by checking for
+linting errors locally:
+
+.. code:: sh
+
+   $ make lint
```

Comment:
just a note that it looks like there's a typo in the `Makefile` at https://github.com/ethereum/web3.py/blob/master/Makefile#L25. It should be `tox -e lint` - it seems to still work fine locally with this typo - but maybe it's worth correcting in this pr - or I can add in a new pr. lmk

Reply:
Odd that it does get parsed appropriately. Will fix now.